source/frontend/editors/RarcEditor.cpp

```diff
@@ -0,0 +1,1035 @@
+        if (!folder)
+          return;
+        editor->InsertFolder(*folder, node.id);
+      }
+      if (ImGui::MenuItem("Create Folder..."_j)) {
+        m_context_modal_state = ModalState::M_OPENING;
+        m_name_input = "";
+        m_original_name = "";
+        editor->SetParentNode(node.id);
+      }
+    }
+    if (ImGui::MenuItem("Delete"_j)) {
+      editor->DeleteNodes({node});
+    }
+    if (ImGui::MenuItem("Rename..."_j)) {
+      m_context_modal_state = ModalState::M_OPENING;
+      m_name_input = node.name;
+      m_original_name = node.name;
+      m_focused_node = node;
+    }
+    if (ImGui::MenuItem("Extract..."_j)) {
+      auto folder = rsl::OpenFolder("Extract Node"_j, "");
+      if (!folder)
+        return;
+      editor->ExtractNodeTo(node, *folder);
+    }
+    if (ImGui::MenuItem("Replace..."_j)) {
+      std::vector<std::string> filters = {"All Files", "*"};
+      if (node.is_folder()) {
+        auto folder = rsl::OpenFolder("Replace With"_j, "");
+        if (!folder)
+          return;
+        editor->ReplaceNodeWith(node, *folder);
+      } else {
+        std::string extension = node.name.substr(node.name.find(".") + 1);
+        if (s_extension_map.contains(extension)) {
+          auto& info = s_extension_map[extension];
+          filters.push_back(info.m_category);
+          filters.push_back(std::format("*.{}", extension));
+        }
+        auto file = rsl::OpenOneFile("Replace With"_j, "", filters);
+        if (!file)
+          return;
+        editor->ReplaceNodeWith(node, *file);
+      }
+    }
+  }
+}
+
+void RarcEditorPropertyGrid::DrawNameModal(RarcEditor* editor) {
+  std::string dialog_name =
+      m_original_name != "" ? "Rename Node"_j : "New Folder"_j;
+
+  if (m_context_modal_state == ModalState::M_OPENING) {
+    ImGui::OpenPopup(dialog_name.c_str());
+  }
+
+  auto flags = ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoMove |
+               ImGuiWindowFlags_AlwaysAutoResize;
+  bool open = true;
+
+  if (ImGui::BeginPopupModal(dialog_name.c_str(), &open, flags)) {
+    RSL_DEFER(ImGui::EndPopup());
+
+    m_context_modal_state = ModalState::M_OPEN;
+
+    char input_buf[128]{};
+    snprintf(input_buf, sizeof(input_buf), "%s", m_name_input.c_str());
+
+    if (ImGui::InputText("Name", input_buf, sizeof(input_buf),
+                         ImGuiInputTextFlags_CharsNoBlank |
+                             ImGuiInputTextFlags_AutoSelectAll)) {
+      m_name_input = input_buf;
+    }
+
+    bool is_invalid = m_name_input == "" || m_name_input == m_original_name;
+
+    if (is_invalid) {
+      // Then, we get the bounding box of the last item
+      ImRect bb = {ImGui::GetItemRectMin(), ImGui::GetItemRectMax()};
+      bb.Max.x -= 36; // Adjust for the name label
+
+      // Then we use the ImGui draw list to draw a custom border
+      ImDrawList* draw_list = ImGui::GetWindowDrawList();
+      draw_list->AddRect(bb.Min, bb.Max, IM_COL32(255, 0, 0, 255)); // RGBA
+    }
```

Comment:
This is clever and I'd love to use this in other places--can you pull this out into a helper function for now?

Reply:
Indeed.